c7n/resources/iam.py

```diff
@@ -37,14 +37,236 @@ class Role(QueryResourceManager):
 
 @resources.register('iam-user')
 class User(QueryResourceManager):
-
     resource_type = 'aws.iam.user'
 
 
+@resources.register('iam-policy')
+class Policy(QueryResourceManager):
+    resource_type = 'aws.iam.policy'
+
+
+@resources.register('iam-profile')
+class InstanceProfile(QueryResourceManager):
+    resource_type = 'aws.iam.instance-profile'
+
+
+@resources.register('iam-certificate')
+class ServerCertificate(QueryResourceManager):
+    resource_type = 'aws.iam.server-certificate'
+
+
+class IamRoleUsage(Filter):
+
+    def get_service_roles(self):
+        results = []
+        for result in self.scan_lambda_roles():
+            if result not in results:
+                results.append(result)
+        for result in self.scan_ecs_roles():
+            if result not in results:
+                results.append(result)
+        for result in self.scan_asg_roles():
+            if result not in results:
+                results.append(result)
+        for result in self.scan_ec2_roles():
+            if result not in results:
+                results.append(result)
+        return results
+
+    def get_instance_profiles(self):
+        results = []
+        for result in self.scan_asg_roles():
+            if result not in results:
+                results.append(result)
+        for result in self.scan_ec2_roles():
+            if result not in results:
+                results.append(result)
+        return results
+
+    def scan_lambda_roles(self):
+        results = []
+        from c7n.resources.awslambda import AWSLambda
+        manager = AWSLambda(self.manager.ctx, {})
+        for l in manager.resources():
+            if 'Role' not in l:
+                continue
+            results.append(l['Role'])
+        return results
+
+    def scan_ecs_roles(self):
+        results = []
+        client = local_session(self.manager.session_factory).client('ecs')
+        for cluster in client.describe_clusters()['clusters']:
+            svcs = client.list_services(cluster=cluster)['serviceArns']
+            for svc in client.describe_services(cluster=cluster,services=svcs)['services']:
+                if 'roleArn' not in svc:
+                    continue
+                results.append(svc['roleArn'])
+        return results
+
+    def scan_asg_roles(self):
+        from c7n.resources.asg import LaunchConfig
+        manager = LaunchConfig(self.manager.ctx, {
+            'resource': 'launch-config'})
+
+        results = []
+        for g in manager.resources():
```

Comment:
a list comprehension can be used here to good effect, also can remove stray commented code.